src/native/minipal/cpufeatures.c

```diff
@@ -441,35 +450,90 @@ int minipal_getcpufeatures(void)
     return result;
 }
 
-// Detect if the current process is running under the Apple Rosetta x64 emulator
-bool minipal_detect_rosetta(void)
+static bool GetCpuBrand(char* brand, size_t bufferSize)
 {
 #if defined(HOST_AMD64) || defined(HOST_X86)
     // Check for CPU brand indicating emulation
     int regs[4];
-    char brand[49];
 
     // Get the maximum value for extended function CPUID info
     __cpuid(regs, (int)0x80000000);
     if ((unsigned int)regs[0] < 0x80000004)
     {
-        return false; // Extended CPUID not supported
+        brand[0] = '\0'; // Extended CPUID not supported, return empty string or handle error
+        return false;
     }
 
-    // Retrieve the CPU brand string
+    // Retrieve the CPU brand string directly into the caller-provided buffer
     for (unsigned int i = 0x80000002; i <= 0x80000004; ++i)
     {
         __cpuid(regs, (int)i);
         memcpy(brand + (i - 0x80000002) * sizeof(regs), regs, sizeof(regs));
     }
-    brand[sizeof(brand) - 1] = '\0';
+
+    brand[bufferSize - 1] = '\0';
+
+    return true;
+#else
+    (void)brand;
+    (void)bufferSize;
+    return false;
+#endif // HOST_AMD64 || HOST_X86
+}
+
+// Detect if the current process is running under the Apple Rosetta x64 emulator
+bool minipal_detect_rosetta(void)
+{
+    char brand[49];
+
+    // Check if CPU brand indicates emulation
+    return GetCpuBrand(brand, sizeof(brand)) && (strstr(brand, "VirtualApple") != NULL);
+}
+
+#if !defined(HOST_WINDOWS)
+
+// Detect if the current process is running under QEMU
+bool minipal_detect_qemu(void)
+{
+    char brand[49];
 
     // Check if CPU brand indicates emulation
-    if (strstr(brand, "VirtualApple") != NULL)
+    if (GetCpuBrand(brand, sizeof(brand)) && strstr(brand, "QEMU") != NULL)
     {
         return true;
     }
-#endif // HOST_AMD64 || HOST_X86
+
+    // Check for process name of PID 1 indicating emulation
+    char cmdline[256];
+    FILE *cmdline_file = fopen("/proc/1/cmdline", "r");
+    if (cmdline_file != NULL)
+    {
+        fgets(cmdline, sizeof(cmdline), cmdline_file);
+        fclose(cmdline_file);
+
+        if (strstr(cmdline, "qemu-") != NULL)
+        {
+            return true;
+        }
+    }
+
+    // Check for process-level emulation using ps command
+    FILE *ps_output = popen("ps -e -o comm=", "r");
+    if (ps_output != NULL)
+    {
+        char process_name[256];
+        while (fgets(process_name, sizeof(process_name), ps_output) != NULL)
+        {
+            if (strstr(process_name, "qemu") != NULL)
+            {
+                pclose(ps_output);
+                return true;
+            }
+        }
+        pclose(ps_output);
+    }
 
     return false;
```

Comment:
Unfortunately in our case none of above checks work because we use https://en.wikipedia.org/wiki/Binfmt_misc on docker and qemu-riscv64-static command is not called explicitly.

Reply:
Do you think you can come up with some way to detect QEMU for purpose of test suppression in #104094? It can be even something as simple as checking `DOTNET_RUNNING_UNDER_QEMU` environment variable that you set for your runs.

Reply:
@am11: 
> Can you share the dockerfile? I'm not checking for qemu-x-static anywhere.

I will paste it in next comment, it's quite large. Regarding checks - if I understand _minipal_detect_qemu_ correctly, you search 'qemu' prefix in ps output. And It works fine but only when I run binary with qemu-x-static. 

@jkotas:
> It can be even something as simple as checking DOTNET_RUNNING_UNDER_QEMU environment variable that you set for your runs.

This is something I started looking for, maybe Qemu sets something in environment that we could use.

Reply:
@jkotas: 
After reading more about topic here and there I think there is no 100% reliable way of detection, it's always some set of heuristics. 
Qemu detection could be "solved" by mentioned simple _DOTNET_RUNNING_UNDER_QEMU_-like workaround. Another a bit more generic way like reading environment and searching for any "QEMU" variable would work for us since we set at least _QEMU_LD_PREFIX_. Unfortunately Qemu doesn't guarantee to provide anything in environment by default which complicates things a bit. 
The second option would be going back to original, ugly approach with pretending that Ubuntu + RISC-V is RISC-V Qemu. However as already noticed it doesn't help with issues that are not RISC-V specific. 
I don't have strong opinion of any of those options, they all have pros and cons.

Reply:
For https://github.com/dotnet/runtime/pull/104094, the important thing is that we have the QEMU detection method for the purpose of test suppression centralized in one place and that we are not overloading the meaning of Ubuntu + RISC to mean QEMU. 

For the implementation of the QEMU detection method itself, I do not have strong opinion about the heuristics to use, as long as it works for your environment and that it is only used for test suppression. Either pre-existing or new environment variable is fine. I expect that we are going to tweak that as needed.